### PR TITLE
(1) add classroom section component for teacher dashboard

### PR DESCRIPTION
**Pull request by jfdoming**
## Notion ticket link

[Teacher Dashboard page](https://www.notion.so/uwblueprintexecs/Teacher-Dashboard-page-3d48a79f5d9342f5afc7f6d56fe230d5)

## Implementation description

image

- Extract hook for class-card data
- Add `thin` variant to classroom card
- Add classroom section for dashboard page
- Support limit option on backend

## Steps to test
1. View component library page.
2. Check that things visually match the design.

## What should reviewers focus on?
* 

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: frontend/src/components/teacher/student-management/classroom-summary/ClassroomCard.tsx [hunk 1]
```diff
@@ -22,6 +22,7 @@ import ClassroomPopover from "./ClassroomPopover";
 
 interface ClassroomCardProps {
   id: string;
+  thin?: boolean;
```
Comment (carissa-tang) [inline frontend/src/components/teacher/student-management/classroom-summary/ClassroomCard.tsx]: maybe rename to like `dashboardView`? smth a bit more encompassing of the two changes

**Review by carissa-tang — COMMENTED** [review]
image

it looks like we want to display all classes actually, with pagination and max 6 per page. but i do think having both pagination and "view all" is a bit crowded and excessive. any thoughts? @joyce-shi 

also, if i'm understanding correctly, looks like we want to sort the classes by the most recently assigned / upcoming test session. i'd be fine with just sorting by `updated` time for the sake of code simplicity and realistically most updates will be of adding test sessions

**Comment by jfdoming** [comment]
TODO filter out archived classrooms - maybe update error message too

**Comment by jfdoming** [comment]
TODO update title to "Classrooms"

**Comment by jfdoming** [comment]
Thoughts:
- No pagination? Probably not
- View all with <= 6 classrooms?

File: backend/services/implementations/classService.ts [hunk 2]
```diff
@@ -106,10 +108,18 @@ class ClassService implements IClassService {
...
   ): Promise<Array<ClassResponseDTO>> {
     let classes: Class[];
     try {
-      classes = await MgClass.find({ teacher: { $eq: teacherId } });
+      const query = MgClass.find({
+        teacher: { $eq: teacherId },
+      });
+      applyQueryOptions(query, queryOptions);
+      if (queryOptions?.excludeArchived) {
+        query.where({ isActive: { $in: [true, undefined] } });
```
Comment (carissa-tang) [inline backend/services/implementations/classService.ts:120]: `undefined` wouldn't be needed if our data was correct right? cause `isActive` is default set to `true` on class creation

File: backend/utilities/generalUtils.ts [hunk 3]
```diff
@@ -47,3 +48,19 @@ export const mapDocumentToDTO = <T extends Document>(document: T) =>
 
 export const mapDocumentsToDTOs = <T extends Document>(documents: Array<T>) =>
   documents.map(mapDocumentToDTO);
+
+const pickQueryOptions = (queryOptions: MongooseQueryOptions): QueryOptions => {
+  const { limit, skip, sort } = queryOptions;
+  return { limit, skip, sort };
+};
```
Comment (carissa-tang) [inline backend/utilities/generalUtils.ts]:
```suggestion
const pickQueryOptions = ({ limit, skip, sort }: MongooseQueryOptions): QueryOptions => ({ limit, skip, sort });
```
maybe more elegant?

File: backend/services/interfaces/classService.ts [hunk 4]
```diff
@@ -59,10 +63,14 @@ export interface IClassService {
   /**
    * This method retrieves the classes associated with given teacher id.
    * @param id teacher id
+   * @param queryOptions optional options for the query including sorting and pagination
```
Comment (carissa-tang) [inline backend/services/interfaces/classService.ts]:
```suggestion
   * @param queryOptions optional options for the query including sorting, filtering and pagination
```

File: frontend/src/components/teacher/dashboard/ClassroomsSection.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    <Box flex="1" w="100%">
+      {loading && (
+        <Center flex="1" margin="15%" mt="calc(15% - 2rem)">
+          <LoadingState />
+        </Center>
+      )}
+      {error && (
+        <Box height="100%" mt={2}>
+          <ErrorState />
+        </Box>
```
Comment (carissa-tang) [inline frontend/src/components/teacher/dashboard/ClassroomsSection.tsx:33]: we should prob standardize the styling for load / error states with theme. 

@joyce-shi do you think its in scope for MVP to do? we can create ticket for it or merge with https://www.notion.so/uwblueprintexecs/Add-loading-state-to-login-pages-and-logout-button-29db3f493a4d412084eddd60af59a107?pvs=4 if its not too much work

**Review by carissa-tang — COMMENTED** [review]
left a few suggestions!

File: backend/services/implementations/classService.ts (hunk 2 above)
Comment (jfdoming) [reply backend/services/implementations/classService.ts:120]: Yes, but right now our database contains a lot of classes where this field is unset. I figured it was probably better to just add the check than do a big migration

Comment (carissa-tang) [reply backend/services/implementations/classService.ts:120]: hmm, ig we prob wouldn't want that check in prod tho? you can add it for testing purposes or at least put TODO to remove

Comment (carissa-tang) [reply backend/services/implementations/classService.ts:120]: idk up to you, maybe better for backwards compatibility purposes

Comment (jfdoming) [reply backend/services/implementations/classService.ts:120]: yeah, my thinking is basically that it doesn't hurt to have this check in prod, and it makes the code overall safer, so unless it starts causing a performance problem I think it's best to keep it

**Review by carissa-tang — APPROVED** [review]
i think you might need to rebase/merge with staging but lgtm!

File: frontend/src/components/teacher/dashboard/ClassroomsSection.tsx (hunk 5 above)
Comment (joyce-shi) [reply frontend/src/components/teacher/dashboard/ClassroomsSection.tsx:33]: yeah let’s merge it with one of the existing load tickets. not _super critical_ but definitely very nice to have I think

Comment (joyce-shi) [reply frontend/src/components/teacher/dashboard/ClassroomsSection.tsx:33]: plus we should try and get this merged soon :D I think it’s been sitting for a while :-/

Comment (jfdoming) [reply frontend/src/components/teacher/dashboard/ClassroomsSection.tsx:33]: indeed it has been merged! added to the linked ticket but ff to move

**Pull request merged**